json_rpc/src/filters.rs

```diff
@@ -0,0 +1,211 @@
+        trace!("{:?}", rejection);
+        message = rejection.to_string();
+        code = StatusCode::UNSUPPORTED_MEDIA_TYPE;
+    } else if let Some(rejection) = error.find::<MissingContentTypeHeader>() {
+        trace!("{:?}", rejection);
+        message = rejection.to_string();
+        code = StatusCode::BAD_REQUEST;
+    } else if let Some(rejection) = error.find::<MissingId>() {
+        trace!("{:?}", rejection);
+        message = rejection.to_string();
+        code = StatusCode::BAD_REQUEST;
+    } else if let Some(rejection) = error.find::<BodyTooLarge>() {
+        trace!("{:?}", rejection);
+        message = rejection.to_string();
+        code = StatusCode::PAYLOAD_TOO_LARGE;
+    } else if error.is_not_found() {
+        trace!("{:?}", error);
+        message = "Path not found".to_string();
+        code = StatusCode::NOT_FOUND;
+    } else if let Some(rejection) = error.find::<reject::MethodNotAllowed>() {
+        trace!("{:?}", rejection);
+        message = rejection.to_string();
+        code = StatusCode::METHOD_NOT_ALLOWED;
+    } else if let Some(rejection) = error.find::<reject::InvalidHeader>() {
+        trace!("{:?}", rejection);
+        message = rejection.to_string();
+        code = StatusCode::BAD_REQUEST;
+    } else if let Some(rejection) = error.find::<reject::MissingHeader>() {
+        trace!("{:?}", rejection);
+        message = rejection.to_string();
+        code = StatusCode::BAD_REQUEST;
+    } else if let Some(rejection) = error.find::<reject::InvalidQuery>() {
+        trace!("{:?}", rejection);
+        message = rejection.to_string();
+        code = StatusCode::BAD_REQUEST;
+    } else if let Some(rejection) = error.find::<reject::MissingCookie>() {
+        trace!("{:?}", rejection);
+        message = rejection.to_string();
+        code = StatusCode::BAD_REQUEST;
+    } else if let Some(rejection) = error.find::<reject::LengthRequired>() {
+        trace!("{:?}", rejection);
+        message = rejection.to_string();
+        code = StatusCode::LENGTH_REQUIRED;
+    } else if let Some(rejection) = error.find::<reject::PayloadTooLarge>() {
+        trace!("{:?}", rejection);
+        message = rejection.to_string();
+        code = StatusCode::PAYLOAD_TOO_LARGE;
+    } else if let Some(rejection) = error.find::<reject::UnsupportedMediaType>() {
+        trace!("{:?}", rejection);
+        message = rejection.to_string();
+        code = StatusCode::UNSUPPORTED_MEDIA_TYPE;
+    } else if let Some(rejection) = error.find::<warp::filters::cors::CorsForbidden>() {
+        trace!("{:?}", rejection);
+        message = rejection.to_string();
+        code = StatusCode::FORBIDDEN;
+    } else {
+        // We should handle all rejection types before this.
+        warn!(?error, "unhandled warp rejection in json-rpc server");
+        message = format!("Internal server error: unhandled rejection: {:?}", error);
+        code = StatusCode::INTERNAL_SERVER_ERROR;
+    }
```

Comment:
Looks like a lot of repetition - can't we add a common trait (if we can't use `warp`'s `Reject`) that will include `code` and `message`?

Reply:
Unfortunately I think this about our best option of we want to use the `to_string()` method and extract the right error code.  `Rejection` doesn't provide any way to extract that without converting to a specific different rejection type, so we'd end up having to implement the new trait for all the various types above.  Seems like we'd just be moving the repetitive code from one place to another.

Reply:
I still prefer to put responsibility on the developer that adds a new `Reject` variant. I think it provides better type-safety as, for example here, one could easily omit a new variant and end up in the `else` clause. My rule of thumb is to avoid `if`s if possible 😝